google/cloud/bigquery_storage_v1/writer.py

```diff
@@ -316,64 +173,46 @@ def close(self, reason: Optional[Exception] = None):
                 an "intentional" shutdown. This is passed to the callbacks
                 specified via :meth:`add_close_callback`.
         """
-        self._shutdown(reason=reason)
-
-    def _shutdown(self, reason: Optional[Exception] = None):
-        """Run the actual shutdown sequence (stop the stream and all helper threads).
-
-        Args:
-            reason:
-                The reason to close the stream. If ``None``, this is
-                considered an "intentional" shutdown.
+        self._connection.close(reason=reason)
+        for callback in self._close_callbacks:
+            callback(self, reason)
+
+        self._closed = True
+
+    def _renew_connection(self, reason: Optional[Exception] = None) -> None:
+        """Helper function that is called when the RPC connection is closed
+        without recovery. It first creates a new Connection instance in an
+        atomic manner, and then cleans up the failed connection. Note that a
+        new RPC connection is not established by instantiating _Connection,
+        but only when `send()` is called for the first time.
         """
-        with self._closing:
-            if self._closed:
-                return
-
-            # Stop consuming messages.
-            if self.is_active:
-                _LOGGER.debug("Stopping consumer.")
-                self._consumer.stop()
-            self._consumer = None
-
-            if self._rpc is not None:
-                self._rpc.close()
-            self._closed = True
-            _LOGGER.debug("Finished stopping manager.")
-
-            # We know that no new items will be added to the queue because
-            # we've marked the stream as closed.
-            while not self._futures_queue.empty():
-                # Mark each future as failed. Since the consumer thread has
-                # stopped (or at least is attempting to stop), we won't get
-                # response callbacks to populate the remaining futures.
-                future = self._futures_queue.get_nowait()
-                if reason is None:
-                    exc = bqstorage_exceptions.StreamClosedError(
-                        "Stream closed before receiving a response."
-                    )
-                else:
-                    exc = reason
-                future.set_exception(exc)
-
-            for callback in self._close_callbacks:
-                callback(self, reason)
-
-    def _on_rpc_done(self, future):
-        """Triggered whenever the underlying RPC terminates without recovery.
+        # Creates a new Connection instance, but doesn't establish a new RPC
+        # connection. New connection is only started when `send()` is called
+        # again, in order to save resource if the stream is idle. This action
+        # is atomic.
+        with self._thread_lock:
+            self._closed_connection = self._connection
```

Comment:
should this just be a local variable rather than on self?  What happens if you get two _renew_connections in succession before shutdown can get called?

Reply:
Good point! It would be a lot better to use a local variable.

Reply:
done

Reply:
> What happens if you get two _renew_connections in succession before shutdown can get called?

On a second thought, I feel like we had better use `self._closed_connection` as a dictionary, and pop the connection once it is successfully shut down. This way we can control the max number of connections in the process of shutting down, and thus control the number of "retries" in a reasonably short amount of time.

Reply:
After some offline discussion, we agree that we will use local variable here instead, and will think of a more holistic way to manage how many times the connection has been retried.